swap/tests/testutils/mod.rs

```diff
@@ -391,21 +404,163 @@ where
     testfn(test).await
 }
 
+fn random_prefix() -> String {
+    use rand::Rng;
+    const CHARSET: &[u8] = b"abcdefghijklmnopqrstuvwxyz";
+    const LEN: usize = 8;
+    let mut rng = rand::thread_rng();
+
+    let prefix: String = (0..LEN)
+        .map(|_| {
+            let idx = rng.gen_range(0, CHARSET.len());
+            CHARSET[idx] as char
+        })
+        .collect();
+    prefix
+}
+
 async fn init_containers(cli: &Cli) -> (Monero, Containers<'_>) {
-    let bitcoind = Bitcoind::new(&cli, "0.19.1").unwrap();
-    let _ = bitcoind.init(5).await;
+    let bitcoind_name = format!("{}{}", random_prefix(), "bitcoind");
+    let volume = random_prefix();
+    let (bitcoind, bitcoind_url) =
+        init_bitcoind_container(&cli, volume.clone(), bitcoind_name.clone(), volume.clone())
+            .await
+            .expect("could not init bitcoind");
+    let electrs = init_electrs_container(&cli, volume.clone(), bitcoind_name, volume).await;
+    let (monero, monerods) = init_monero_container(&cli).await;
+    (monero, Containers {
+        bitcoind_url,
+        bitcoind,
+        monerods,
+        electrs,
+    })
+}
+
+async fn init_bitcoind_container(
+    cli: &Cli,
+    volume: String,
+    name: String,
+    network: String,
+) -> Result<(Container<'_, Cli, bitcoind::Bitcoind>, Url)> {
+    let image = bitcoind::Bitcoind::default()
+        .with_mapped_port(Port {
+            local: 7041,
+            internal: 18443,
+        })
+        .with_mapped_port(Port {
+            local: 7042,
+            internal: 18886,
+        })
+        .with_volume(volume)
+        .with_tag("0.19.1");
+
+    let run_args = RunArgs::default().with_name(name).with_network(network);
+
+    let docker = cli.run_with_args(image, run_args);
+
+    let bitcoind_url = {
+        let input = format!("http://{}:{}@localhost:{}", "admin", "123", 7041);
+        Url::parse(&input).unwrap()
+    };
+
+    init_bitcoind(bitcoind_url.clone(), 5).await?;
+
+    Ok((docker, bitcoind_url.clone()))
+}
+
+pub async fn init_electrs_container(
+    cli: &Cli,
+    volume: String,
+    bitcoind_container_name: String,
+    network: String,
+) -> Container<'_, Cli, electrs::Electrs> {
+    let bitcoind_rpc_addr = format!("{}:18443", bitcoind_container_name);
+    let image = electrs::Electrs::default()
+        .with_mapped_port(Port {
+            local: 3012,
+            internal: 3002,
+        })
+        .with_mapped_port(Port {
+            local: 60401,
+            internal: 60401,
+        })
+        .with_volume(volume)
+        .with_daemon_rpc_addr(bitcoind_rpc_addr)
+        .with_tag("latest");
+
+    let run_args = RunArgs::default().with_network(network);
+
+    cli.run_with_args(image, run_args)
+}
```

Comment:
Passing the `Cli` into this function is weird. You are only using it in the last line of the function which means this function might as well just "create" the `Image` instance and return it, something like:

```rust
pub async fn new_electrs_image(
    volume: String,
    bitcoind_container_name: String,
) -> electrs::Electrs {
    let bitcoind_rpc_addr = format!("{}:18443", bitcoind_container_name);
    
    electrs::Electrs::default()
        .with_mapped_port(Port {
            local: 3012,
            internal: 3002,
        })
        .with_mapped_port(Port {
            local: 60401,
            internal: 60401,
        })
        .with_volume(volume)
        .with_daemon_rpc_addr(bitcoind_rpc_addr)
        .with_tag("latest")
}
```

That would simplify the signature quite a bit without majorly impacting the call-site.

Reply:
We let docker map the ports now. To do that we need to create the `Container` which requires the `&Cli`.